### PR TITLE
Implementation of the Ruby enumerable method

### PR DESCRIPTION
**Pull request by babs257**
This is a feature branch of my implementation of Ruby Enumerable Methods

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job :+1: 
You have done a splendid job implementing the requirements for this project. Keep up.
I tested your methods against some edge cases and some have failed. Please look at my comments and implement those changes.

### Martin Afani

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
+module Enumerable
+    def my_each
+        for item in self
+            yield(item)
+        end
+        self
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
returns an Enumerator if no block is given. You can do this `def my_each return to_enum unless block_given?`

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+   
+    def my_each_with_index
+        i = 0
+        while i < self.length
+         yield(self[i], i)
+        i += 1
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)).
Returns an enumerator if no block is given. Can be implemented as above.

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    self
+    end
+
+    def my_select
+        new_array = []
+        self.my_each do |item|
+          new_array.push(item) if yield(item)
+        end
+        new_array
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)).
returns an enumerator if no block is given

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    def my_all?
+        test = true
+        self.my_each do |item|
+          test = false if !yield(item)
+          break
+        end
+        return test
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
- [ ]  when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    
+    def my_any?
+        test = false
+        self.my_each do |item|
+          test = true if yield(item)
+          break
+        end
+        return test
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ]  when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    
+    def my_none?
+        test = true
+        self.my_each do |item|
+          test = false if yield(item)
+          break
+        end
+        return test
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true only if none of the collection members is true
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    
+    def my_count
+        total = 0
+        if block_given?
+          self.my_each {|item| total += 1 if yield(item)}
+        else
+          total = self.count
+        end
+        return total
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
 - [ ] counts the number of items in enum that are equal to item if an argument is given

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        new_array = []
+        self.my_each do |value|
+          if my_proc.nil?
+            new_array.push(my_proc.call(value))
+          else
+            new_array.push(yield(value))
+          end
+        end
+        return new_array
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an Enumerator if no block is given. do something like `def my_map return to_enum unless block_given?`

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    
+    def my_inject(memo=self[0])
+        self.my_each_with_index do |value, index|
+          memo = yield(memo, value) if index > 0
+        end
+        return memo
+      end
+    end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] combines all elements of enum by applying a binary operation, specified by a block
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method

File: enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    def my_all?
+      return to_enum unless block_given?
+
+        test = true
+        self.my_each do |item|
+          test = false if !yield(item)
+          break
+        end
+        return test
+    end
```
Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]** 
* It should return the same thing as Ruby `all?` method. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from the Odin).
* When no block or argument is given returns true when none of the collection members are false or nil.
Screenshot from 2019-12-30 13-14-40
* When a class is passed as an argument returns true if all of the collection is a member of such class.
Screenshot from 2019-12-30 13-21-55
* When a Regex is passed as an argument returns true if all of the collection matches the Regex.
```
Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)
     
     ArgumentError:
       wrong number of arguments (given 1, expected 0)
```
* When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern.
Screenshot from 2019-12-30 13-32-03

File: enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    def my_any?
+      return to_enum unless block_given?
+
+        test = false
+        self.my_each do |item|
+          test = true if yield(item)
+          break
+        end
+        return test
+    end
```
Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]**
* Returns the same thing as Ruby's `any?` enumerable method ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from the Odin).
* When no block or argument is given returns true if at least one of the collection is not false or nil.
Screenshot from 2019-12-30 13-37-18
* When a class is passed as an argument returns true if at least one of the collection is a member of such class.
Screenshot from 2019-12-30 13-38-43
* When a Regex is passed as an argument returns false if none of the collection matches the Regex.
Screenshot from 2019-12-30 13-41-10
* When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern.
Screenshot from 2019-12-30 13-42-35

File: enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    def my_none?
+      return to_enum unless block_given?
+      
+        test = true
+        self.my_each do |item|
+          test = false if yield(item)
+          break
+        end
+        return test
+    end
```
Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]** 
* Returns the same value as Ruby `none?` enumerable method. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from the Odin).
* When no block or argument is given returns true only if none of the collection members is true.
Screenshot from 2019-12-30 13-46-42
* When a class is passed as an argument returns true if none of the collection is a member of such class.
Screenshot from 2019-12-30 13-48-22
* When a Regex is passed as an argument returns true only if none of the collection matches the Regex.
Screenshot from 2019-12-30 13-49-45
* When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern.
Screenshot from 2019-12-30 13-52-10

File: enumerable.rb [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    
+    def my_count
+        total = 0
+        if block_given?
+          self.my_each {|item| total += 1 if yield(item)}
+        else
+          total = self.count
+        end
+        return total
+    end
```
Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]**
* Returns the same value as Ruby's `count` enumerable method. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from the Odin).
* Counts the number of items in enum that are equal to item if an argument is given.
Screenshot from 2019-12-30 13-55-57

File: enumerable.rb [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        new_array = []
+        self.my_each do |value|
+          if my_proc.nil?
+            new_array.push(my_proc.call(value))
+          else
+            new_array.push(yield(value))
+          end
+        end
+        return new_array
+    end
```
Comment (wathigo) [inline enumerable.rb]: * Returns the same value as ruby's `map` enumerable method. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from Odin).
* Returns an Enumerator if no block is given.
Screenshot from 2019-12-30 13-59-27

File: enumerable.rb [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    
+    def my_inject(memo=self[0])
+        self.my_each_with_index do |value, index|
+          memo = yield(memo, value) if index > 0
+        end
+        return memo
+      end
+    end
```
Comment (wathigo) [inline enumerable.rb]: **[REQUIRED]**
* Returns the same value as ruby's `inject` enumerable method. ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from Odin).
* Combines all elements of enum by applying a binary operation, specified by a block.
Screenshot from 2019-12-30 14-05-37
* When a symbol is specified combines each element of the collection by applying the symbol as a named method.
Screenshot from 2019-12-30 14-09-02

**Review by wathigo — CHANGES_REQUESTED** [review]
## Changes Required 
### Hi @babs257  :wave:  
We have a few issues we need to fix before moving on.
See the comments :point_down: 

**Happy Coding :computer: 
Simon Wathigo**

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Hey @babs257 
I see you have made some changes but still, your methods fail against most edge cases. Please take a look at the comments left by my colleague @wathigo. Please submit another review when done.

### Martin Afani

**Review by imhta — APPROVED** [review]
# Project approved 🎉 

### Hello  @babs257  , 👋 You have done a great job in this project 👏 Every test passed successfully.




----
### Happy coding 😄 
----
**Pon Muthu Selvam N**
__nponmuthuselvam@gmail.com__

**Pull request merged**